File: 자료구조/prim_알고리즘.c
// 2026-06-03

// prim_알고리즘.c
#include <stdio.h>

#define MAX 4
#define INF 9999

typedef struct GraphType {
    int n;
    int adj_mat[MAX][MAX];
} GraphType;

void init(GraphType* g) {
    g->n = MAX;

    for (int i = 0; i < MAX; i++) {
        for (int j = 0; j < MAX; j++) {
            g->adj_mat[i][j] = 0;
        }
    }
}

void insert_edge(GraphType* g, int u, int v, int weight) {
    g->adj_mat[u][v] = weight;
    g->adj_mat[v][u] = weight;
}

void prim(GraphType* g, int start) {
    int selected[MAX] = {0};
    int total_weight = 0;

    selected[start] = 1;

    printf("Prim MST 간선\n");

    for (int count = 0; count < g->n - 1; count++) {
        int min = INF;
        int u = -1;
        int v = -1;

        for (int i = 0; i < g->n; i++) {
            if (selected[i]) {
                for (int j = 0; j < g->n; j++) {
                    if (!selected[j] && g->adj_mat[i][j] != 0) {
                        if (g->adj_mat[i][j] < min) {
                            min = g->adj_mat[i][j];
                            u = i;
                            v = j;
                        }
                    }
                }
            }
        }

        selected[v] = 1;
        total_weight += min;

        printf("%d - %d : %d\n", u, v, min);
    }

    printf("총 가중치 : %d\n", total_weight);
}

int main() {
    GraphType g;

    init(&g);

    insert_edge(&g, 0, 1, 10);
    insert_edge(&g, 0, 2, 6);
    insert_edge(&g, 0, 3, 5);
    insert_edge(&g, 1, 3, 15);
    insert_edge(&g, 2, 3, 4);

    prim(&g, 0);

    return 0;
}

File: 자료구조/prim_알고리즘.py
# 2026-06-03

# prim_알고리즘.py
MAX = 4
INF = 9999

class GraphType:
    def __init__(self):
        self.n = MAX
        self.adj_mat = [[0 for _ in range(MAX)] for _ in range(MAX)]

def insert_edge(g, u, v, weight):

    g.adj_mat[u][v] = weight
    g.adj_mat[v][u] = weight

def prim(g, start):

    selected = [False] * g.n
    selected[start] = True

    total_weight = 0

    print("Prim MST 간선")

    for _ in range(g.n - 1):

        min_weight = INF
        u = -1
        v = -1

        for i in range(g.n):

            if selected[i]:

                for j in range(g.n):

                    if (not selected[j] and
                        g.adj_mat[i][j] != 0):

                        if g.adj_mat[i][j] < min_weight:

                            min_weight = g.adj_mat[i][j]
                            u = i
                            v = j

        selected[v] = True
        total_weight += min_weight

        print(f"{u} - {v} : {min_weight}")

    print("총 가중치 :", total_weight)

g = GraphType()

insert_edge(g, 0, 1, 10)
insert_edge(g, 0, 2, 6)
insert_edge(g, 0, 3, 5)
insert_edge(g, 1, 3, 15)
insert_edge(g, 2, 3, 4)

prim(g, 0)